internal/check/check.go

```diff
@@ -0,0 +1,153 @@
+package check
+
+import (
+	"bytes"
+	"context"
+	"encoding/json"
+	"fmt"
+	"io"
+	"net/http"
+	"runtime"
+	"strconv"
+	"sync"
+	"text/tabwriter"
+	"time"
+
+	"golang.org/x/sync/semaphore"
+)
+
+type RunFunc func(ctx context.Context) (string, error)
+
+type Check struct {
+	Name        string
+	Description string
+
+	Run RunFunc
+
+	LastRun           time.Time
+	CachedCheckErr    error
+	CachedCheckOutput string
+}
+
+func (c *Check) MarshalJSON() ([]byte, error) {
+	return json.Marshal(struct {
+		Name              string    `json:"name"`
+		Description       string    `json:"description"`
+		LastRun           time.Time `json:"last_run"`
+		CachedCheckErr    string    `json:"cached_check_err"`
+		CachedCheckOutput string    `json:"cached_check_output"`
+	}{
+		Name:              c.Name,
+		Description:       c.Description,
+		LastRun:           c.LastRun,
+		CachedCheckErr:    errString(c.CachedCheckErr),
+		CachedCheckOutput: c.CachedCheckOutput,
+	})
+}
+
+type HealthChecker struct {
+	mu     sync.Mutex
+	Checks []Check
+}
+
+func (c *HealthChecker) MarshalJSON() ([]byte, error) {
+	c.mu.Lock()
+	defer c.mu.Unlock()
+
+	return json.Marshal(c.Checks)
+}
+
+func (c *HealthChecker) Run() {
+	for {
+		time.Sleep(60 * time.Second)
+		c.doRun()
+	}
+}
+
+func (c *HealthChecker) doRun() {
+	ctx, cancel := context.WithTimeout(context.Background(), time.Second*10)
+	defer cancel()
+
+	sem := semaphore.NewWeighted(int64(runtime.GOMAXPROCS(0)))
+	wg := sync.WaitGroup{}
+
+	for i, check := range c.Checks {
+		err := sem.Acquire(ctx, 1)
+		if err != nil {
+			return
+		}
+		wg.Add(1)
+		go func(i int, check Check) {
+			defer sem.Release(1)
+			defer wg.Done()
+
+			check.CachedCheckOutput, check.CachedCheckErr = check.Run(ctx)
+			check.LastRun = time.Now().UTC()
+
+			c.mu.Lock()
+			c.Checks[i] = check
+			c.mu.Unlock()
+		}(i, check)
+	}
+	wg.Wait()
+}
+
+func (c *HealthChecker) Print(w http.ResponseWriter) {
```

Comment:
Nice. I like the idea of each check having a debug page on the service itself, in addition to an endpoint that the frontend service can scrape.